### ISSUE TITLE
Prefetch batches earlier in Ruby and JS drivers

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#4795]
As of RethinkDB 2.1.2, our drivers send a CONTINUE on a cursor, then wait for the response, parse it, and send up to one additional CONTINUE request immediately for prefetching purposes.

This works fine as long as the bottleneck on the client is in the application consuming the data from the cursor. However if the bottleneck is in the parsing step, then prefetching the next batch only after parsing the previous response doesn't do much.

This can have a significant impact on the throughput at which one can read from a cursor for workloads such as the one described in https://github.com/rethinkdb/rethinkdb/issues/4569 .

For the Python driver, this was fixed in https://github.com/rethinkdb/rethinkdb/commit/e94e1656cb12e9f2793a14361dd71afbf7849983 and will ship together with RethinkDB 2.1.3.
The new driver now sends the next CONTINUE request as soon as the previous reply has arrived, *before* parsing the response (this means that we sometimes send another CONTINUE after the cursor has been exhausted, but that's fine).

We should also optimize the Ruby and JavaScript drivers in this respect (and in the Java driver too! :-) ).

**Comment by neumino** [comment]
I don't know about the Ruby driver but from what I remember the JS driver already does that. It prefetches the next batch as soon as it's processing the last batch.

**Comment by danielmewes** [comment]
I don't believe it does.
If you look at `Connection._data`, it does the following:
```
[...]
# Since we have enough data, we slice it out of the
# buffer, and parse it as JSON.
responseBuffer = @buffer.slice(12, responseLength + 12)
response = JSON.parse(responseBuffer)

# Now it's off to `_processResponse` where the data is
# converted to a format the user will be able to work
# with, error responses are detected etc.
@_processResponse response, token
```

`_processResponse` calls `_addResponse` on the right cursor, which in turn calls `_promptNext` which does the prefetching.

So `JSON.parse` is done before we start fetching the next batch from what I can tell.

**Comment by neumino** [comment]
Oh, I misunderstood the issue.

In https://github.com/rethinkdb/rethinkdb/commit/e94e1656cb12e9f2793a14361dd71afbf7849983#diff-d73ba4e9a072dff9a19fe841ca493c8fL335, the driver doesn't take a peek at the response. So in the case when you get a `SUCCESS_SEQUENCE`, doesn't the driver fire an extra `CONTINUE` query?

**Comment by Tryneus** [comment]
@neumino, yes, but we keep track of the cursor until the outstanding request is resolved - and if we get responses for an ended cursor (due to a `close()`, an error, or a `SUCCESS_SEQUENCE`), the response will be discarded.  It may put a little extra load on the server, but it should be rather negligible.

**Comment by deontologician** [comment]
Technically, since the json response is always serialized the same way, we could peek at the beginning of the response and check if its success sequence without parsing the entire thing. I believe the `r` key shows up first

**Comment by deontologician** [comment]
I don't think we should do that but it's totally possible

**Comment by mlucy** [comment]
@Tryneus -- wait, so we keep around information about cursors after we've already sent `SUCCESS_SEQUENCE`?  How long do we keep them around for?

**Comment by danielmewes** [comment]
@mlucy I think we only do that as long as we still have an outstanding request for the cursor. @Tryneus will know for sure.

**Comment by Tryneus** [comment]
I discussed this with @mlucy in person and he agreed that this behavior should be fine.

We only hold onto the information about the cursors in the driver - the server discards it as soon as a terminal response has been sent for a cursor.  It is a little weird that our drivers will send an extra request for a completed cursor (which will always get an error), but that is probably better than holding onto cursors on the server after they've completed so we could send a success response of some kind.
